File: Applied Course/4.Problem Solving/4.Problems on Stacks and Queues/6.The Celebrity Problem.py
import sys
mat=[[0,0,0,0,1],
     [0,0,0,0,1],
     [0,0,0,0,1],
     [0,0,0,0,1],
     [0,0,0,0,0]]
class node: 

    def __init__(self, info): 
        self.info = info  
        self.next = None 


class Stack: 

    def __init__(self): 
        self.top = None
        
    
    def isEmpty(self):
        if self.top is None:
            return True
        return False
    
    def push(self,data):
        self.temp=node(data)
        if self.temp is None:
            print("Stack overflow")
            return
        self.temp.next=self.top
        self.top=self.temp
    
    def pop(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        d=self.top.info
        self.top=self.top.next    
        return d
    
    def peek(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        d=self.top.info
        return d
    def display(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        self.p=self.top
        while self.p is not None:
            print(self.p.info)
            self.p=self.p.next
            
    def getSize(self):
        self.p=self.top
        count=0
        while self.p is not None:
            count+=1
            self.p=self.p.next
            
def haveAquantiance(a,b):
    return mat[a][b]

def findCeleb(n):
    stack=Stack()
    for i in range(n):
        stack.push(i)
        
    temp1=stack.pop()
    temp2=stack.pop()
    
    while(stack.getSize()>1):
        if(haveAquantiance(temp1,temp2)):
            temp1=stack.pop()
        else:
            temp2=stack.pop()
    celeb=stack.pop()
    if(haveAquantiance(celeb,temp2)):
        celeb=temp2
    if(haveAquantiance(celeb,temp1)):
        celeb=temp1
    
    for i in range(n):
        if(i!=celeb and haveAquantiance(celeb,i) and haveAquantiance(i,celeb)!=1):
            return -1
    return celeb
    
if __name__=='__main__':
    val=findCeleb(5)
    if(val>0):
        print("Celebrity found {}".format(val))
    else:
        print("Celebrity not found")

File: Applied Course/4.Problem Solving/4.Problems on Stacks and Queues/6.The Celebrity Problem.c
#include <stdio.h>
#define MAX 20
#define N 10

int stack[20];
int top=-1;
int mat[N][N]={{0,0,0,0,1},
			   {0,0,0,0,1},
			   {0,0,0,0,1},
		       {0,0,0,0,1},
		       {0,0,0,0,0}};

// Function to check both persons knows each other or not
int HaveAcquaintance(int a, int b)
{
	return mat[a][b];
}


//Function to check the stack is empty or not
int isEmpty()
{
	return ((top) == -1);
}

// Function for push the element into stack
void push(int data)
{
	if(top==MAX){
		printf("Stack overflow");
		return;
	}
	stack[++top] = data;
}

// Function to pop an element from the stack
int pop()
{
	if(isEmpty()){
		printf("Stack Empty");
		return -1;
	}
	else{
		int value = stack[top--];
		return value;
	}
}

// Function to check size of the stack
int size()
{
	return top+1;
}

// Returns -1 if celebrity is not present. If present, returns index(id)
//(value from 0 to n-1). 

int findCeleb(int n)
{
	int celeb;
	// push everybody to stack
	for(int i; i<n; i++)
		push(i);
	
	// get top2 elements
	int temp1 = pop();
	int temp2 = pop();

	//Find potential celebrity
	while(size()>1)
	{
		if(HaveAcquaintance(temp1,temp2))
			temp1 = pop();
		else
			temp2 = pop();
	}
	
	celeb = pop();

	if(HaveAcquaintance(celeb,temp2))
		celeb=temp2;
	if(HaveAcquaintance(celeb,temp1))
		celeb=temp1;

	// Check if Celeb is actually a celebrity or not
	for (int i = 0; i < n; i++) 
    {  
    	// If any person doesn't know 'p' or 'p' doesn't know any person, return -1
        if ( (i != celeb) && (HaveAcquaintance(celeb, i) || !HaveAcquaintance(i, celeb)) ) 
            return -1; 
    } 
    return celeb;
}

int main()
{
	int n = 5;
	int value = findCeleb(n);
	if(value>0)
		printf("the celebrity is %d\n",value);
	else{
		printf("celebrity not found\n");
	}
	return 0;
}